### PR TITLE
Implement swipe on Sidenav

### LINKED ISSUE #275: Swipe on Sidenav
I got it to work, still need to make it more flexible and think about a solution for multiple sidenavs. But first, do you want it?

ezgif com-447b40bb5e

### PR DESCRIPTION
**Pull request by igor-ribeiro**
Fix #275

### CODE REVIEW

**Comment by igor-ribeiro** [comment]
I will change to Mouse Events [http://caniuse.com/#feat=touch]

**Comment by igor-ribeiro** [comment]
Actually, it looks like there are no problems on mobile support, so I guess we can leave it like that.

File: docs/src/pages/components/Sidenav.vue [hunk 1]
```diff
@@ -7,6 +7,36 @@
...
+              <md-table-row>
+                <md-table-head>Name</md-table-head>
+                <md-table-head>Type</md-table-head>
+                <md-table-head>Description</md-table-head>
+              </md-table-row>
+            </md-table-header>
+
+            <md-table-body>
+              <md-table-row>
+                <md-table-cell>swipeable</md-table-cell>
```
Comment (marcosmoura) [inline docs/src/pages/components/Sidenav.vue]: The attributes should be prefixed with "md-" :)

File: docs/src/pages/components/Sidenav.vue [hunk 2]
```diff
@@ -7,6 +7,36 @@
...
+                <md-table-cell><code>Boolean</code></md-table-cell>
+                <md-table-cell>Enable the swipe functionality. Default <code>false</code></md-table-cell>
+              </md-table-row>
+
+              <md-table-row>
+                <md-table-cell>swipe-threshold</md-table-cell>
```
Comment (marcosmoura) [inline docs/src/pages/components/Sidenav.vue]: Same here.

File: docs/src/pages/components/Sidenav.vue [hunk 3]
```diff
@@ -7,6 +7,36 @@
...
+                <md-table-cell><code>Number</code></md-table-cell>
+                <md-table-cell>Set the initial threshold for the swipe when it's closed. Default <code>15</code></md-table-cell>
+              </md-table-row>
+
+              <md-table-row>
+                <md-table-cell>swipe-distance</md-table-cell>
```
Comment (marcosmoura) [inline docs/src/pages/components/Sidenav.vue]: And here.

File: docs/src/pages/components/Sidenav.vue [hunk 4]
```diff
@@ -102,7 +132,7 @@
                 <p>Open console to see the events</p>
               </div>
 
-              <md-sidenav class="md-left" ref="leftSidenav" @open="open('Left')" @close="close('Left')">
+              <md-sidenav class="md-left" ref="leftSidenav" @open="open('Left')" @close="close('Left')" swipeable>
```
Comment (marcosmoura) [inline docs/src/pages/components/Sidenav.vue]: Same here too.

File: src/components/mdSidenav/mdSidenav.vue [hunk 5]
```diff
@@ -45,7 +56,71 @@
         } else {
           this.open();
         }
+      },
+      handleTouchStart(event) {
+        const positionX = event.touches[0].clientX - this.mountedRect.left;
+        const positionY = event.touches[0].clientY - this.mountedRect.top;
+
+        if (positionX < 0 || positionX > this.mountedRect.left + this.mountedRect.width) {
```
Comment (marcosmoura) [inline src/components/mdSidenav/mdSidenav.vue]: Maybe will be good to create three methods with the conditions inside those if statements, and then merge then to only one if, like that:
 
``` js
if (this.methodOne() || this.methodTwo() || this.methodThree()) {
  return;
}
```

File: src/components/mdSidenav/mdSidenav.vue [hunk 6]
```diff
@@ -45,7 +56,71 @@
...
+
+      this.mountedRect = this.$refs.backdrop.$el.getBoundingClientRect();
+      this.initialTouchPosition = null;
+      this.canMove = false;
+
+      document.addEventListener('touchstart', this.handleTouchStart);
+      document.addEventListener('touchend', this.handleTouchEnd);
+      document.addEventListener('touchmove', this.handleTouchMove);
+    },
+    destroyed() {
```
Comment (marcosmoura) [inline src/components/mdSidenav/mdSidenav.vue]: Instead of destroyed, move this to beforeDestroy

**Comment by marcosmoura** [comment]
GREAT job ever here. Just created some comments to make it more compliant with the code style and standards.

File: src/components/mdSidenav/mdSidenav.vue (hunk 5 above)
Comment (igor-ribeiro) [inline src/components/mdSidenav/mdSidenav.vue]: Is this OK for you?
```js
if (!this.isHorizontallyInside(positionX) || !this.isVerticallyInside(positionY) || !this.isFromStartWhenClosed(positionX)) {
  return;
}
```

**Comment by igor-ribeiro** [comment]
The only problem is that we cannot swipe from right to left to open. I can work on it if you want.

**Comment by marcosmoura** [comment]
@igor-ribeiro I think, for now, we don't need to develop this. I will create some components to the the "app-shell". And then will be easier to create ths behaviour. I will take a deep look at this code and I get back to you. Thanks a lot for this PR.

**Comment by fergardi** [comment]
Any idea why this suddenly stopped working for me? Was running fine in my Cordova app but no more swipe effect... :(

**Comment by igor-ribeiro** [comment]
@fergardi Can you detail a little more? I work with someone that is using it and the swipe is working. Try to adjust the props `md-swipe-distance` and `md-swipe-threshold`.

**Comment by fergardi** [comment]
I added it but no luck yet. The milestone for this was changing from default cordova webview to cordova crosswalk webview plugin (latest chromium canary). Before that it worked like a charm. Now thanks to that plugin it seems to load and render really really fast in comparison, specially the images and scrolling behaviour, but in exchange I lost the swipe events (and probably other events) in the process. Not working anymore. At all. Maybe has to do with something related to the new @click.native events? Dunno. For now, I'll keep the xwalk plugin as it speeds up my app like a LOT, but would love to have back the swipe events. Could you add this plugin to a cordova project and test your voodoo magic against it?

And btw, could you point me in the right direction to add your md-swipeable functionality by default to my new component in  #549? Would be REALLY cool. Like... mandatory.

Thanks for your effort.

**Comment by fergardi** [comment]
Never mind, It somehow fixed itself. Maybe was an outdated plugin messing touch events. Btw I added your functionality to my new slider component!

**Comment by Linzdigr** [comment]
Would love to have a simple way to get the right sidenav working with swipe! (instead of the default left to right). Nice job anyway.

**Pull request merged**